src/core/lb/particle_coupling.cpp

```diff
@@ -172,45 +183,109 @@ Utils::Vector3d ParticleCoupling::get_noise_term(Particle const &p) const {
   return m_noise_pref_wo_gamma * Utils::hadamard_product(sqrt(gamma), noise);
 }
 
-void ParticleCoupling::kernel(Particle &p) {
-  auto const agrid = m_lb.get_agrid();
```

Comment:
I've been looking at this funcito for a while now. 
The complexity arises mainly, as in some cases forces are added in more than one place for a single particle.

I wonder if it cannot be simplified by passing over the vector of particle pointers twice.
Precondition: only particles within the local domain +halo (agrid/2) should be in the vector

1. Get all interpolated velocities at particles' folded position. This yields exactly one velocity per particle.
2. Populate arrays with coupling positions and coupling forces. Coupling position and coupling force arrays will again have the same length, but there can be more coupling positions/forces than particles.

    * For nromal particles, the coupling forces comes from the difference of particle and lb velocity. Coupling positions are teh particle position plus, in some caes, images shifted by +-box_l. 
    * For swimmers, the force comes from the particles'swimming properites. Again, forces are applied at the particle osition, plus positions shifted by +- box_l, if within agrid/2 of the boudnary

Or am I missing sth?

Reply:
offline discussion: postpone this refactoring to the coding day